dropshot/tests/test_tls.rs

```diff
@@ -12,52 +12,86 @@ use std::convert::TryFrom;
 
+use crate::common::generate_tls_key;
+
 /// See rustls::client::ServerCertVerifier::verify_server_cert for argument
 /// meanings
-type VerifyCertFn = Box<
+type VerifyCertFn<'a> = Box<
     dyn Fn(
-            &rustls::Certificate,
-            &[rustls::Certificate],
-            &rustls::ServerName,
-            &mut dyn Iterator<Item = &[u8]>,
+            &rustls::pki_types::CertificateDer,
+            &[rustls::pki_types::CertificateDer],
+            &rustls::pki_types::ServerName,
             &[u8],
-            SystemTime,
-        ) -> Result<rustls::client::ServerCertVerified, rustls::Error>
+            rustls::pki_types::UnixTime,
+        )
+            -> Result<rustls::client::danger::ServerCertVerified, rustls::Error>
         + Send
-        + Sync,
+        + Sync
+        + 'a,
 >;
 
-struct CertificateVerifier(VerifyCertFn);
+struct CertificateVerifier<'a>(VerifyCertFn<'a>);
+
+impl<'a> std::fmt::Debug for CertificateVerifier<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("CertificateVerifier... with some function?")
+    }
+}
 
-impl rustls::client::ServerCertVerifier for CertificateVerifier {
+impl<'a> rustls::client::danger::ServerCertVerifier
+    for CertificateVerifier<'a>
+{
     fn verify_server_cert(
         &self,
-        end_entity: &rustls::Certificate,
-        intermediates: &[rustls::Certificate],
-        server_name: &rustls::ServerName,
-        scts: &mut dyn Iterator<Item = &[u8]>,
+        end_entity: &rustls::pki_types::CertificateDer<'_>,
+        intermediates: &[rustls::pki_types::CertificateDer<'_>],
+        server_name: &rustls::pki_types::ServerName<'_>,
         ocsp_response: &[u8],
-        now: SystemTime,
-    ) -> Result<rustls::client::ServerCertVerified, rustls::Error> {
-        self.0(end_entity, intermediates, server_name, scts, ocsp_response, now)
+        now: rustls::pki_types::UnixTime,
+    ) -> Result<rustls::client::danger::ServerCertVerified, rustls::Error> {
+        self.0(end_entity, intermediates, server_name, ocsp_response, now)
+    }
+
+    fn verify_tls12_signature(
+        &self,
+        _message: &[u8],
+        _cert: &rustls::pki_types::CertificateDer<'_>,
+        _dss: &rustls::DigitallySignedStruct,
+    ) -> Result<rustls::client::danger::HandshakeSignatureValid, rustls::Error>
+    {
+        Ok(rustls::client::danger::HandshakeSignatureValid::assertion())
+    }
+
+    fn verify_tls13_signature(
+        &self,
+        _message: &[u8],
+        _cert: &rustls::pki_types::CertificateDer<'_>,
+        _dss: &rustls::DigitallySignedStruct,
+    ) -> Result<rustls::client::danger::HandshakeSignatureValid, rustls::Error>
+    {
+        Ok(rustls::client::danger::HandshakeSignatureValid::assertion())
+    }
+
+    fn supported_verify_schemes(&self) -> Vec<rustls::SignatureScheme> {
+        // Default algorithm from rcgen
+        vec![rustls::SignatureScheme::ECDSA_NISTP256_SHA256]
```

Comment:
What's the deal with these implementations?

Reply:
I wasn't 100% sure on this, but according to https://docs.rs/rustls/latest/rustls/client/danger/trait.ServerCertVerifier.html#tymethod.verify_tls12_signature

> cert has already been validated by [ServerCertVerifier::verify_server_cert](https://docs.rs/rustls/latest/rustls/client/danger/trait.ServerCertVerifier.html#tymethod.verify_server_cert).

So it seemed like waving the signature through preserved the existing behavior. I'm happy to add a comment to that effect; also happy to discuss if this doesn't seem like the right policy.